ckanext/canada/tabulate.py

```diff
@@ -0,0 +1,311 @@
+"""
+Contains subclasses of Tabulator Stream and CSVParser and CSV Dialect.
+
+Xloader and Validation plugins use these classes
+to use static dialects with Tabulator.
+
+The CSVParser is different from others, and needs some extra mangling.
+
+Note: script cannot be named tabulator! (hence tabulate)
+"""
+from tabulator.stream import Stream
+from tabulator.parsers.csv import CSVParser
+from tabulator.config import CSV_SAMPLE_LINES
+from tabulator.exceptions import TabulatorException, EncodingError, SourceError
+import six
+import json
+import csv
+from csv import Dialect
+from _csv import Dialect as _Dialect
+
+from ckan.plugins.toolkit import config, _
+
+from logging import getLogger, Filter
+
+#FIXME: the chances of a user using quad pipes as a
+#       CSV dialect value is low, but never zero...
+ERROR_SPLITTER = '||||'
+
+
+class LimitStreamLogging(Filter):
+    logged_messages = []
+    def filter(self, record):
+        current_log = (record.module, record.levelno, record.msg)
+        if current_log not in self.logged_messages:
+            self.logged_messages.append(current_log)
```

Comment:
this will grow without bounds and searching a list is O(n). You could use a set to make the search O(1) but the growth growth will still be a problem.

Can we store just the last message or last n messages to solve the problem?